pageserver/src/tenant.rs

```diff
@@ -1824,6 +1824,11 @@ impl Tenant {
         pitr: Duration,
         ctx: &RequestContext,
     ) -> anyhow::Result<GcResult> {
+        // Don't start doing work during shutdown
+        if let TenantState::Stopping { .. } = self.current_state() {
+            return Ok(GcResult::default());
+        }
+
         // there is a global allowed_error for this
         anyhow::ensure!(
             self.is_active(),
```

Comment:
Isn't there now repetition with the ensure in both cases?

We do not start the background loops before we are active, next possible steps are broken and stopping.

I am also surprised how did you get to merge this without conflicts since I did make the CompactionError visible in the #4938.

Reply:
I meant repeatition between:

```
if let TenantState::Stopping { .. } = self.current_state()
```

and

```
anyhow::ensure!(
    self.is_active(),
```

And how the stopping OR broken can be the only next state.

Perhaps the suggestion I was working towards is should we instead have:

```patch
-if let TenantState::Stopping { .. } = self.current_state() {
+if !self.is_active() {
     return Ok(GcResult::default());
 }
 
-// there is a global allowed_error for this
-anyhow::ensure!(
-    self.is_active()
-);
```

(did not check how the ensure! block would correctly end).

Reply:
Ah, I see.  The reason I didn't do that is because this function is called from two places: in the compaction loop I would always want to drop out with Ok(()) if we have no work, but when called via the API (e.g. in a test), I would prefer to give the caller an error if they have tried to compact prematurely during activation.

Reply:
Today I found out that the `Tenant::compaction_iteration` is not called from tests. I see that GC of course is.. I'll do what I proposed in #5881 for that so I can use the better error type.

Reply:
> prefer to give the caller an error if they have tried to compact prematurely during activation

Related: there is a PR I intend to merge from external contributor I can never find the number of, but it would add waiting until tenants have finished loading by default to all tests so this reason would go away for gc as well.

It never existed for compaction though, as `Tenant::compaction_iteration` was never exposed like `Tenant::gc` is.